### ISSUE TITLE
Echo at the end of each recording.

### ISSUE THREAD

**Issue by marcom48** [keithito/tacotron#133]
Hi there,

I'm doing some training on my own data, got up to around 120k steps, but at the end of each sentence I synthesise the sentence continually echo. Has anybody come across this and found a solution?

Thanks

**Comment by Quadraaa** [comment]
+1.
I've cut all the silence from the end (even sacrificing some sound sometimes) and am using attention from https://github.com/keithito/tacotron/blob/tacotron2-work-in-progress/models/attention.py as @keithito suggested, but it doesn't help.

**Comment by pgkhang** [comment]
Yeah, i'm having this problem too. The audio keeps repeating the last couple of words.
[eval-69000-2.zip](https://github.com/keithito/tacotron/files/1845784/eval-69000-2.zip)

**Comment by 1413394** [comment]
I'm having same problem. The synthesized audio keep echoing. How can I trim it?

**Comment by toannhu** [comment]
What's about the alignment? IMHO. I think tacotron has still tried to learn how to stop the sequence by the "\n" characters. Maybe your dataset is not large enough to get Tacotron detect whether the end of the speech?

**Comment by Quadraaa** [comment]
@NTT123 Has adding silence solved the problem with echo in your case? I also tried adding 1sec of silence at the end of each training sample, bit it didn't work for me.

**Comment by NTT123** [comment]
@Quadraaa It works for my dataset (7 hours Vietnamese voice). Adding 1 sec silence forces the network to learn to generate a long silence after the end of stream symbol `~`.  

You should listen to `logs/tacotron/step-audio.wav` files. If the network trained correctly, it should be able to generate 1 sec silences at the end of the audio.

**Comment by Quadraaa** [comment]
@NTT123 Thanks for the info. 

Maybe the problem is that my data is not of a very high quality. The length is about 15 hours. If I add 1sec silence the echo effect gets less (although it still exists), but the model starts to generate noise instead. So if I cut all the silence, I get echo with very little noise sometimes, and when I cut all the silence and then add 1 sec of pure silence, I get echo + noise. Going to try a different datasetof a larger size. I also noticed that the speaker in the recording I am using usually makes very long pauses between words. It may be contributing to the problem, I suppouse. Usually the generation stops mistakenly after a comma.

I listened to `logs/tacotron/step-audio.wav` files, the silence at the end is generated.

**Comment by NTT123** [comment]
@Quadraaa, I also though that your problem is the long silences in your training samples. I personally used `pydub` to spit the audio files into audio segments of no longer 500 ms silences.

**Comment by Quadraaa** [comment]
I tried a larger dataset (40h), trimmed the silences inside the audiosamples to be no longer than 500ms, replaced sounds of a volume lower than -60dB with pure silence, added 1 sec of pure silence to the end of each audiosample. Echo got much less, but in some cases it still ocures.  
Have no idea, what else can I do. Was thinking of modifying the loss function so that the model gets penalized heavily for producing the sequence longer than the ground truth sequence, but not shure if it will help because at the training phase the decoder gets ground truth as it's inputs and therefore has no problems with echo.

**Comment by Quadraaa** [comment]
@rafaelvalle 
Sorry, can't say anything about it, because in my case the attention was learnt both with and without removing silences longer than 500ms.

**Comment by rafaelvalle** [comment]
For the people having trouble, can you post the distribution of words per second of your dataset ? I think this is quite important

**Comment by 1413394** [comment]
@rafaelvalle 
Mine is 3.6 words per second (141715 words divided by 39278 seconds) but i omitted comma and dot

**Comment by Quadraaa** [comment]
@rafaelvalle 
I am having troubles too, but anyway:
![words_per_second](https://user-images.githubusercontent.com/12547433/38318364-ce7d0654-3837-11e8-859e-0f34791c1803.png)

I suppose, it may be important to filter the dataset for outliers based on words per second and seconds per word. These are all language-dependent, of course. The distribution of length of the audiosamples may be important too, but according to my experience it is better to have a larger dataset with skewed distribution (in my case there are many short samples and the longer the audiosample length the less examples I have) than a normally distributed dataset created by undersampling.

**Comment by rafaelvalle** [comment]
These are statistics on LJ Speech Dataset, which @keithito uses in this repo.
Please try use samples such that you mimic LJ's speech rate and let us know how it goes!
![screen shot 2018-04-04 at 9 32 14 am](https://user-images.githubusercontent.com/2302607/38321008-18add692-37eb-11e8-9fcb-d5895ee1257e.png)

**Comment by rafaelvalle** [comment]
@1413394 Could you train on your date with 3.6 words per second?

**Comment by 1413394** [comment]
@rafaelvalle  Yes i trained 45k steps with tacotron model using Location Sensitive Attention.  I tried to train with normal bahdanau attention but the synthesized audio keeps echoing, so i tried the location sensitive attention as @keithito suggested but the echo still persists. Here 's my audio synthesized from the model and the alignment at 45k steps. It speaks the text correctly but keeps echoing at the end.
![step-45000-align](https://user-images.githubusercontent.com/37022661/38323108-8e021e82-3866-11e8-9686-c242dab62fae.png)
[sample.zip](https://github.com/keithito/tacotron/files/1876830/sample.zip)

**Comment by rafaelvalle** [comment]
@1413394 That's probably because the decoder is unsure when to stop decoding. Can you share your dataset ?

**Comment by 1413394** [comment]
@rafaelvalle I'm not masking the padding frames. I created my dataset by segmenting some audiobooks. I'll send you an email.

**Comment by NTT123** [comment]
IIRC, in google's papers, there is a subnetwork which outputs an end of audio stream signal (perhaps, a sigmoid output neuron). It could be helpful for solving the echoing problem.

**Comment by rafaelvalle** [comment]
@NTT123 Yes, Taco2 has a gating layer and we're checking if it will fix the "echoing" problem.

**Comment by jlin816** [comment]
Interestingly enough, I'm trying to synthesize subword utterances (e.g. "sa" or "fu") and the echoing is also happening for me. Not sure if this is evidence for/against it being related to word rate.

**Comment by rafaelvalle** [comment]
@jlin816 That's probably because it doesn't have the gate. Can you also share your dataset?

**Comment by jlin816** [comment]
Sure, here are the synthesized samples after 78.9k iterations. Admittedly my task is a little unconventional -- I'm trying to synthesize with phonemic/anatomical parameters rather than text, but it has been working surprisingly well aside from the echo.
[eval-78900.zip](https://github.com/keithito/tacotron/files/1882004/eval-78900.zip)

**Comment by begeekmyfriend** [comment]
In my many tests, I have found that the more steps it trains, the larger probability the ending echoes would happen. I am afraid there should be something to be adjusted in the attention model. @keithito

**Comment by keithito** [comment]
Oh, that's awesome! I'll definitely try it out. :-)

**Comment by hyzhan** [comment]
@rafaelvalle What parameters are there related to the distribution of words per second of your dataset ?

**Comment by tuong-olli** [comment]
@keithito I don't know why tacotron model don't need padding in synthesize.py?
